src/ValueObjects/LatestReceiptInfo.php

```diff
@@ -0,0 +1,374 @@
+     * This field is only present for upgrade transactions.
+     * @var bool|string|null
+     */
+    private $isUpgraded;
+
+    /**
+     * The offer-reference name of the subscription offer code that the customer redeemed.
+     * @see https://developer.apple.com/documentation/appstorereceipts/offer_code_ref_name?changes=latest_minor
+     * @var string|null
+     */
+    private $offerCodeRefName;
+
+    /**
+     * The time of the original app purchase, in UNIX epoch time format, in milliseconds.
+     * @var int|null
+     */
+    private $originalPurchaseDate;
+
+    /**
+     * The transaction identifier of the original purchase.
+     * @see https://developer.apple.com/documentation/appstorereceipts/original_transaction_id?changes=latest_minor
+     * @var string
+     */
+    private $originalTransactionId;
+
+    /**
+     * The unique identifier of the product purchased.
+     * @var string
+     */
+    private $productId;
+
+    /**
+     * The identifier of the subscription offer redeemed by the user.
+     * @see https://developer.apple.com/documentation/appstorereceipts/promotional_offer_id?changes=latest_minor
+     * @var string|null
+     */
+    private $promotionalOfferId;
+
+    /**
+     * The time the App Store charged the user’s account for a purchased or restored product
+     * @var int|null
+     */
+    private $purchaseDate;
+
+    /**
+     * The number of consumable products purchased.
+     * @var int
+     */
+    private $quantity;
+
+    /**
+     * The identifier of the subscription group to which the subscription belongs.
+     * @see https://developer.apple.com/documentation/storekit/skproduct/2981047-subscriptiongroupidentifier?changes=latest_minor
+     * @var string|null
+     */
+    private $subscriptionGroupIdentifier;
+
+    /**
+     * A unique identifier for purchase events across devices, including subscription-renewal events.
+     * This value is the primary key for identifying subscription purchases.
+     * @var string|null
+     */
+    private $webOrderLineItemId;
+
+    /**
+     * A unique identifier for a transaction such as a purchase, restore, or renewal.
+     * @see https://developer.apple.com/documentation/appstorereceipts/transaction_id?changes=latest_minor
+     * @var string
+     */
+    private $transactionId;
+
+    /**
+     * @param string $originalTransactionId
+     * @param string $productId
+     * @param int $quantity
+     * @param string $transactionId
+     */
+    public function __construct(string $originalTransactionId, string $productId, int $quantity, string $transactionId)
+    {
+        $this->originalTransactionId = $originalTransactionId;
+        $this->productId = $productId;
+        $this->quantity = $quantity;
+        $this->transactionId = $transactionId;
+    }
+
+    /**
+     * @param array $attributes
+     * @return static
+     */
+    public static function fromArray(array $attributes): self
+    {
+        $obj = new self(
```

Comment:
fail-fast via guarding?

Reply:
Seems to me it's the time to read more about fail-fast 😄

Reply:
basically only allow instantiating if all required values are valid and given, otherwise fail. if find this to be a good resource: https://www.martinfowler.com/ieeeSoftware/failFast.pdf